blockdag/reachability.go

```diff
@@ -547,9 +811,276 @@ func (dag *BlockDAG) updateReachability(node *blockNode, selectedParentAnticone
 			return err
 		}
 	}
+
+	// Update the reindex root.
+	// Note that we check for blue score here in order to find out
+	// whether the new node is going to be the virtual's selected
+	// parent. We don't check node == virtual.selectedParent because
+	// at this stage the virtual had not yet been updated.
+	if node.blueScore > dag.SelectedTipBlueScore() {
+		updateStartTime := time.Now()
+		modifiedTreeNodes, err := dag.updateReachabilityReindexRoot(newTreeNode)
+		if err != nil {
+			return err
+		}
+		if len(modifiedTreeNodes) > 0 {
+			updateTimeElapsed := time.Since(updateStartTime)
+			log.Debugf("Reachability reindex root updated to %s. "+
+				"Modified %d tree nodes and took %dms.",
+				dag.reachabilityReindexRoot.blockNode.hash,
+				len(modifiedTreeNodes), updateTimeElapsed.Milliseconds())
+			for modifiedTreeNode := range modifiedTreeNodes {
+				dag.reachabilityStore.setTreeNode(modifiedTreeNode)
+			}
+		}
+	}
+
 	return nil
 }
 
+func (dag *BlockDAG) updateReachabilityReindexRoot(newTreeNode *reachabilityTreeNode) (
+	map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	nextReindexRoot := dag.reachabilityReindexRoot
+	for {
+		candidateReindexRoot, modifiedNodes, found, err := dag.tryMovingReachabilityReindexRoot(nextReindexRoot, newTreeNode)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			break
+		}
+		for node := range modifiedNodes {
+			modifiedTreeNodes[node] = struct{}{}
+		}
+		nextReindexRoot = candidateReindexRoot
+	}
+
+	dag.reachabilityReindexRoot = nextReindexRoot
+	return modifiedTreeNodes, nil
+}
+
+func (dag *BlockDAG) tryMovingReachabilityReindexRoot(
```

Comment:
This is not "try" because it does not imply "do something, if you fail - we're fine". It's a "maybe" at the very least.
But anyway, as mentioned in comment before - if possible and makes sense - better to restructure this whole thing